hazelcast/src/test/java/com/hazelcast/client/AuthenticationLeakTest.java

```diff
@@ -0,0 +1,228 @@
+        authenticateAndAssert(AuthenticationStatus.SERIALIZATION_VERSION_MISMATCH, (byte) (serVersion + 1), false, clusterName);
+    }
+
+    @Test
+    public void testFailedAuthenticationDoesNotLeakInfoClientNotAllowed() throws Exception {
+        // No client is allowed in the cluster
+        getClientEngineImpl(instance).applySelector(ClientSelectors.none());
+        authenticateAndAssert(AuthenticationStatus.NOT_ALLOWED_IN_CLUSTER, serVersion, false, clusterName);
+    }
+
+    private void authenticateAndAssert(AuthenticationStatus status, byte serVersion, boolean useWrongClusterName, String clusterName) throws IOException {
+        InetSocketAddress endpoint = new InetSocketAddress("127.0.0.1", 5701);
+        try (Socket socket = new Socket()) {
+            socket.setReuseAddress(true);
+            socket.connect(endpoint);
+            try (OutputStream os = socket.getOutputStream(); InputStream is = socket.getInputStream()) {
+                os.write(CLIENT_BINARY.getBytes(StandardCharsets.UTF_8));
+                String clientClusterName = useWrongClusterName ? clusterName + 'a' : clusterName;
+                ClientMessage res = authenticate(clientClusterName, serVersion, os, is);
+                ClientAuthenticationCodec.ResponseParameters responseParameters = ClientAuthenticationCodec.decodeResponse(res);
+                assertEquals(status.getId(), responseParameters.status);
+                assertEquals(-1, responseParameters.partitionCount);
+                assertEquals(-1, responseParameters.serializationVersion);
+                assertNull(responseParameters.address);
+                assertNull(responseParameters.memberUuid);
+                assertNull(responseParameters.clusterId);
+            }
+        }
+    }
+
+    private ClientMessage authenticate(String clusterName, byte serVersion, OutputStream os, InputStream is)
+            throws IOException {
+        UUID uuid = new UUID(0, 0);
+        ClientMessage msg = ClientAuthenticationCodec.encodeRequest(clusterName, null, null, uuid, "", serVersion, "", "", new ArrayList<>());
+        writeClientMessage(os, msg);
+        return readResponse(is, ClientAuthenticationCodec.RESPONSE_MESSAGE_TYPE);
+    }
+
+    private void getKeyValue(SerializationService ss, OutputStream os, InputStream is)
+            throws IOException {
+        Data keyData = ss.toData("key");
+        ClientMessage msg = MapGetCodec.encodeRequest("mapName", keyData, 0);
+        msg.setPartitionId(getPartitionId(keyData));
+        writeClientMessage(os, msg);
+        readResponse(is, MapGetCodec.RESPONSE_MESSAGE_TYPE);
+    }
+
+    private int getPartitionId(Data keyData) {
+        int hash = keyData.getPartitionHash();
+        return HashUtil.hashToIndex(hash, 271);
+    }
+
+    private void writeClientMessage(OutputStream os, final ClientMessage clientMessage) throws IOException {
```

Comment:
I think since this code is duplicated 3 times (this, phonehome.TestUtil, ClientMessageProtectionTest) maybe it would be a good idea to put this code into some util class